polling_stations/apps/api/authentication/authentication.py

```diff
@@ -1,19 +1,47 @@
+from django.conf import settings
+from django.contrib.auth.backends import ModelBackend
+from django.contrib.auth.models import AnonymousUser
 from rest_framework.authentication import TokenAuthentication
+from rest_framework.exceptions import AuthenticationFailed
 
+TokenAuthentication
 
-class TokenAuthSupportQueryString(TokenAuthentication):
-    """
-    Extend the TokenAuthentication class to support querystring authentication
-    in the form of "http://www.example.com/?auth_token=<token_key>"
-    """
 
-    def authenticate(self, request):
-        # Check if 'token_auth' is in the request query params.
-        # Give precedence to 'Authorization' header.
-        if (
-            "auth_token" in request.query_params
-            and "HTTP_AUTHORIZATION" not in request.META
-        ):
-            return self.authenticate_credentials(request.query_params.get("auth_token"))
+class ReadOnlyTokenAuthenticatedUser(AnonymousUser):
+    def __init__(self, token):
+        super().__init__()
+        self.token = token
 
-        return super(TokenAuthSupportQueryString, self).authenticate(request)
+    @property
+    def is_authenticated(self):
+        return True
+
+    @property
+    def is_superuser(self):
+        return False
+
+
+class SuperTokenAuthenticatedUser(ReadOnlyTokenAuthenticatedUser):
+    @property
+    def is_superuser(self):
+        return True
+
+
+class HardcodedTokenAuthentication(ModelBackend):
+    def authenticate(self, request, **kwargs):
+        token: str = (
+            request.GET.get("auth_token")
+            or request.META.get("HTTP_AUTHORIZATION")
+            or ""
+        )
+        token = token.removeprefix("Token ")
+        if token in getattr(settings, "READ_ONLY_API_AUTH_TOKENS", []):
+            user = ReadOnlyTokenAuthenticatedUser(token=token)
+            return user, None
+        if token in getattr(settings, "SUPERUSER_API_AUTH_TOKENS", []):
```

Comment:
Ditto this is an array of tokens for WDIV developers/human users who want an API key to try things out?

Reply:
I think the only use case for this at the moment is the Lambda function posting back to this app after upload processing.